consensus/core/src/leader_scoring.rs

```diff
@@ -71,9 +178,160 @@ impl ReputationScores {
     }
 }
 
+pub(crate) struct UnscoredSubdag {
+    pub context: Arc<Context>,
+    pub blocks: BTreeMap<BlockRef, VerifiedBlock>,
+}
+
+impl UnscoredSubdag {
+    pub(crate) fn new(context: Arc<Context>, subdags: &[CommittedSubDag]) -> Self {
+        let blocks = subdags
+            .iter()
+            .flat_map(|subdag| subdag.blocks.iter())
+            .map(|block| (block.reference(), block.clone()))
+            .collect::<BTreeMap<_, _>>();
+
+        assert!(
+            !blocks.is_empty(),
+            "Attempted to create UnscoredSubdag with no blocks"
+        );
+
+        Self { context, blocks }
+    }
+
+    // Skip genesis round as we don't produce leaders for that round.
+    pub(crate) fn get_leader_rounds(&self) -> Vec<Round> {
+        self.blocks
+            .keys()
+            .map(|block_ref| block_ref.round)
+            .filter(|round| *round != 0)
+            .collect::<Vec<_>>()
+    }
```

Comment:
We are just trying to get the rounds in which there may potentially be a leader and we do this by getting all the rounds of blocks that exist in the `unscored_subdag`. Then when we go to calculate scores for a leader because we are looping through the committer we will get the expected leaders within the same round because the leader offset is a part of the committer.

I will use a set so we filter out the duplicate rounds though